### ISSUE TITLE
Optimize Admin to Load Applets with 50 Activities

### ISSUE THREAD

**Issue by WorldImpex** [ChildMindInstitute/mindlogger-admin-OLD-Vue#772]
1. When trying to open this applet in the admin dashboard, the lazy loader is shown and user's page never loads. We need to investigate why this happens 

The goal of this effort is that the users page loads and the data visualization page works for each user. 

https://raw.githubusercontent.com/charlie42/reproschema-builder/mhdb-format/protocols/sc_dd/sc_dd_schema

2. Editing a large applet the first time takes a long time (DBT & EMA). We need to optimize as well as add a lazy loader so the user understands that something is happening

**Comment by WorldImpex** [comment]
Please check this applet https://github.com/ChildMindInstitute/reproschema-builder/blob/master/applets/protocols/test_large_applet/test_large_applet_schema 

1. When trying to add this applet to a device, the user is required to sign out and sign back in for the applet to appear 
2. Trying to edit this applet with the builder does not work and just shows lazy loader

**Comment by WorldImpex** [comment]
[#1722](https://app.zenhub.com/workspaces/mindlogger-5e11094d0c26311588da9626/issues/childmindinstitute/mindlogger-app/1722) is also impacted by an account with more than 4 applets

**Comment by jj105** [comment]
this is link for an applet with 2000+ items.

https://raw.githubusercontent.com/ChildMindInstitute/reproschema-builder/master/applets/protocols/hbn_all/hbn_all_schema

**Comment by sergeysemenuta** [comment]
Checked this on the admin manel for this large applet https://raw.githubusercontent.com/ChildMindInstitute/reproschema-builder/master/applets/protocols/hbn_all/hbn_all_schema

On the calendar it works good, but data viz dashboard are not loaded at all after taking some activities by  users
Account: bughunterqa123@gmail.com // 123456 'hbn all' applet
**Video:**https://www.screencast.com/t/SgPW8N3EomOW

Editing the DBT, NIMH  applets were fixed
Video: https://www.screencast.com/t/GLqYggIDc7

**Note:** Really bad performance from app side for this applet. After scheduling some events in the future time for example 5+ activities it takes a  lot of time to load by tapping the applet card or activity card. Sometimes it takes minutes to load depends on how much activities is in scheduled status
Video: https://www.screencast.com/t/wxWFZLMLT5N
https://admin-staging.mindlogger.org/ (Win 10 // Chrome Version 90.0.4430.212)
ML v 0.16.23
iPhone X // iOS 13.3
Pixel 4XL//Android 11

**Comment by sergeysemenuta** [comment]
Rechecked on the Staging admin panel and on the ML App (api Dev environment)
ML v 0.16.23
iPhone X // iOS 13.3
Pixel 4XL//Android 11

Video: https://www.screencast.com/t/zANvY7aQ2A
Checked on the  Hbn All applet
- Loading the data viz takes about 15 seconds faster than before

- On the mobile app it works faster as well 
23 events are scheduled in the future time and loading the activities screen takes about  3-5 seconds on android and ios platforms
Video:https://www.screencast.com/t/iizcN7ICCuyP

**Comment by sergeysemenuta** [comment]
Failed on the admin side
Loading activities  takes to much time on the calendar. The admin panel becomes unresponsive for some time after switching filters
Account: bughunterqa123@gmail.com // 123456, 'hbn all' applet
Note: Events are scheduled to repeat daily,monthly etc. 

Video: https://www.screencast.com/t/4uk0O7Wm

Win 10 // Chrome 91

**Comment by natalia-muzyka** [comment]
1) Loading activities in the calendar go much faster now, mostly with a delay in 1-2 seconds for each newly displayed (scrolled to) part of the calendar.
Once I've experienced that displaying of one week stuck with loader(for more than 1 minute), it was fixed after switching the calendar view. Couldn't reproduce this anymore. 
There are no such delays while switching the calendar view, the biggest delay is 2-3 seconds after switching to the "Year" view.
Didn't face the unresponsive error.
Video: https://www.screencast.com/t/SYmBpEhJZu
Account:
bughunterqa123@gmail.com // 123456

2) The creating of a new applet with this schema takes more than 10 minutes delay (~15 minutes or more):
https://raw.githubusercontent.com/ChildMindInstitute/reproschema-builder/master/applets/protocols/hbn_all/hbn_all_schema

3) After an attempt to edit an applet hbn_all_schema with 2000+ items:
- at the first time(after logging in) clicking the "Edit" btn the "Your applet is being edited" pop-up is shown
- after clicking the "Edit" btn on the dashboard again the applet edit page isn't opened, and "Edit" btn is blocked for ~5 minutes
When "Edit" btn becomes active the edit applet page loads in 3-5 seconds after clicking the "Edit" button.
After editing and saving the applet the activities in the calendar start displaying incorrect (e.g. format name "qr97_schema"), isn't fixed after clearing cache and re-login.
Reproduced for two accounts:
bughunterqa123@gmail.com // 123456
geria1.test@gmail.com // 7clCm7D5
Video: https://www.screencast.com/t/YMHkadwBz8

4) Loading the DataViz depends on how much new info was received and can take up to 13-15 seconds (after completing the activity with 20items+). Expanding such activity on DataViz can take 2-5 seconds.
Video: https://www.screencast.com/t/WSpr2OFY6

Accounts:
bughunterqa123@gmail.com // 123456
geria1.test@gmail.com // 7clCm7D5

Environment:
Win 10 // Chrome https://admin-staging.mindlogger.org/
ML v0.16.25 (staging) 
Samsung Galaxy S7 Edge SM-G935FD // Android 8.0
iPhone XR // iOS 14.3

**Comment by WorldImpex** [comment]
@jj105 

3) Do you now if all items are already supported in applet builder?

**Comment by jj105** [comment]
@WorldImpex yes, I think all items we have in github are editable via builder.

**Comment by natalia-muzyka** [comment]
@jj105 @WorldImpex 
One new thing mentioned today and created a new ticket for it. 
Infinity loader on Applet Library. The part of applets can't be loaded and added to the cart
https://app.zenhub.com/workspaces/mindlogger-webapp-5e11094d0c26311588da9626/issues/childmindinstitute/mindlogger-applet-library/64

**Comment by WorldImpex** [comment]
@natalia-muzyka 3 was updated to be more informative so the user knows what is going to happen. The rest are okay for mvp

**Comment by jj105** [comment]
@natalia-muzyka for 3, if large applet was edited backend sends email to editor.
by clicking a link in the email, editor will be able to landed on applet builder.

**Comment by natalia-muzyka** [comment]
Verified: if large applet was edited backend sends email to editor.
by clicking a link in the email, editor is able to landed on applet builder.

Noticed: Even when the user is logged in on admin panel, the email link redirects him to the log-in page. After logging in the user is redirected to the appropriate applet builder page. Is this behavior acceptable for now?

Video: https://www.screencast.com/t/yKJNGkkA
Environment:
https://admin-staging.mindlogger.org/
ML & gmail account: geria1.test@gmail.com // 7clCm7D5

@WorldImpex @jj105

**Comment by natalia-muzyka** [comment]
@jj105 
This issue is still reproduced: 
**Even when the user is logged in on admin panel, the email link redirects him to the log-in page. After logging in the user is redirected to the appropriate applet builder page.** After logging in the appropriate builder is opened.
https://www.screencast.com/t/o0cPIoPHmCBl
https://www.screencast.com/t/vLMQGy1lXBi

**Environment:**
https://admin-staging.mindlogger.org/
ML & gmail account: geria1.test@gmail.com // 7clCm7D5

**Comment by natalia-muzyka** [comment]
@jj105 After clicking the link, if the user was previously authorized on the admin panel, the applet builder is opened.

The same behavior(issue) now is for **non-logged in** user.
Video: https://www.screencast.com/t/qit20f0o

@WorldImpex , correct me please if I'm wrong
We should have the next behavior when the user clicks the link:
- if authorized: redirect to an applet builder;
- if non-authorized: redirect to the "Log in" page -> redirect to an applet builder after logging in with valid credentials

**Comment by WorldImpex** [comment]
@natalia-muzyka That is correct

**Comment by WorldImpex** [comment]
@natalia-muzyka Due to technical constraints, regardless of the logged-in state of the user, we are going to require the login

**Comment by natalia-muzyka** [comment]
Verified: if large applet was edited backend sends email to editor.
by clicking a link in the email, the "log in" page is displayed for authorized and non-authorized user.
The editor is able to land on applet builder after logging in.

Environment:
https://admin-staging.mindlogger.org/
ML & gmail account: geria1.test@gmail.com // 7clCm7D5

**Comment by natalia-muzyka** [comment]
@jj105 There is an infinity loader in the applet builder and after refreshing the admin panel the "Edit" button is greyed on the copied: 
- hbn_all applet
- NIMH applet

Video: https://www.screencast.com/t/5uPi5pxPl

Environment:
https://admin-staging.mindlogger.org/
geria1.test@gmail.com	// 7clCm7D5
applet: hbn all (7)

https://admin.mindlogger.org/
loui_ml@protonmail.com // 12345678
applet: NIMH
natalia.muzyka@qatestlab.net // 19911170 (switch to health brain network)
applet: New daisy chain applet (1)

Applet password	Qwe123!!!

**Comment by natalia-muzyka** [comment]
@jj105 

**1)** For **copied** big applets there is still an issue. The user doesn't see a confirmation pop-up, doesn't receive the confirmation mail with the applet link. There is only an infinity loader after an attempt to edit the **copied** applet.
Video: https://www.screencast.com/t/Xk5xnfl20psq
bernard_ml@protonmail.com	// qwe123!!!
applet: NIMH (optimization test)

**2)** The "Edit" button is still greyed on old copied applets
https://admin-staging.mindlogger.org/
geria1.test@gmail.com // 7clCm7D5
applet: hbn all (7), hbn all (8)
Video: https://www.screencast.com/t/bOZ6f7ZP

https://admin-staging.mindlogger.org/
Win 10 // Chrome 96.0.4664.45

**Comment by jj105** [comment]
I think for duplicated applets, we don't need to show confirmation pop-up because they are not connected to github url directly
( applets connected to github url are ones that managers can see refresh button when he hovers applet in the list )

![image.png](https://images.zenhubusercontent.com/5e565f9bd9593e11ebab005a/872d047f-5f78-4460-8852-34293e420cf7)

**Comment by natalia-muzyka** [comment]
@jj105 I'm sorry, I wasn't clear about the pop-up. I mean there should be a pop-up "Your applet is being edited" when a user tries to edit a big applet for the first time, isn't it? And once editing is available user receives the appropriate email.

**Comment by jj105** [comment]
@natalia-muzyka -- I see, yes I think there should be a pop-up and user should receive email when editing is available
I fixed that issue yesterday, I think it's ready to be checked on admin-staging

**Comment by natalia-muzyka** [comment]
testing is blocked by https://github.com/ChildMindInstitute/mindlogger-backend/issues/1165

**Comment by natalia-muzyka** [comment]
Verified as fixed.
#1 Video: https://www.screencast.com/t/8Z1NPGSIqp7
#2 Video: https://www.screencast.com/t/XxevBPH1MMp

https://admin-staging.mindlogger.org/
Win 10 // Chrome 96.0.4664.45 
macOS 10.15 // Chrome 96.0.4664.55 
geria1.test@gmail.com // 7clCm7D5
applet: hbn all (7), hbn all (8)
bernard_ml@protonmail.com	// qwe123!!!
applet: NIMH, NIMH (optimization test)
